### DISCUSSION TITLE
GSoC projects

### DISCUSSION THREAD

**Discussion by pca006132** [elalish/manifold#697] [General]
List of mentors:
1. @elalish 
2. @pca006132 
3. @zalo 

Potential Projects:
- Better mesh ID handling (#678) (@pca006132)
- Add minGap function (#186) (@pca006132)
- Fix vert properties and improve documentation on vert properties (#536 #661)
- SDF batch LevelSet (#655)
- Add more fuzz tests (#667) (@pca006132)
- Mesh UV map generation (#683)
- Improve mesh simplification (#675)

Some are probably too hard for GSoC. We can probably add some js/python projects as well to attract beginners.

**Comment by elalish** [comment]
Great starter list, thanks! We can remove refine to length - I'm about halfway done with that now. A few others I've thought of:

- Add undo, redo, format, and share buttons to ManifoldCAD.org.
- Better make-manifold function with hole filling.
- Refine to precision - make heuristic for number of divisions of each edge based on its tangents.

**Reply by starseeker** [reply]
I'm curious about the idea of a "make-manifold" function with hole filling.  One of the problems we often see in meshes we get from conversions is "almost manifold" meshes that have some tiny gaps or topological flaws in an otherwise closed mesh.  For our use case any sort of localized repair that resulted in a close manifold would probably suffice - is the ability you're describing there intended for such a use case?

**Reply by elalish** [reply]
We actually already have the [Merge](https://manifoldcad.org/docs/html/structmanifold_1_1_mesh_g_l.html#a2078d075bef279041b67769e4197c254) function for that. You can even test it on GLBs here: https://manifoldcad.org/make-manifold, which can save the GLB with our new `EXT_mesh_manifold` extension that ensures these topological flaws don't creep in anymore.

**Reply by zalo** [reply]
I think I'd also be interested in adding a way to make sharing models on `manifoldcad` easier...
Either via hash params in the URL or some stateful server backend?

**Reply by elalish** [reply]
Actually, @pca006132 already implemented just that: https://github.com/elalish/manifold/pull/643 - we just need to hook it up to a button.

**Reply by pca006132** [reply]
Not sure if the manifoldcad.org button project will be too trivial for GSoC. Maybe a good fit for short (75h)?

**Reply by elalish** [reply]
Yeah, and honestly I might do it before summer anyway.

**Reply by pca006132** [reply]
@elalish thinking if it is a good idea to combine the mesh ID thing and improving mesh simplification. They are related, and combining them can free one slot (considering we can have at most 2 slots).

**Reply by elalish** [reply]
Sure, sounds good.

**Comment by pca006132** [comment]
Another project: optimize quickhull implementation. I think it should be possible to look into the code, find places where there are bottlenecks and do something with it. E.g. parallelization.

**Reply by pca006132** [reply]
copying @zalo's comment on zulip:

I believe the VHACD library switched from QuickHull to Julio Jerez (of Newton Physics)'s Convex Hull algorithm because it's much faster and more numerically stable:
https://github.com/kmammou/v-hacd/blob/master/include/VHACD.h#L1988-L3489

**Reply by pca006132** [reply]
@zalo would you like to mentor this? I think you are more knowledgeable than me about this.

**Reply by zalo** [reply]
I have some idea about the pitfalls in common convex hull algorithms that I might be able to provide PR review/supervision for if anyone is interested in tackling speed improvements 👍

**Comment by pca006132** [comment]
@elalish are we using the metallic menger sponge as logo?

**Reply by elalish** [reply]
Indeed - and it looks like @t-paul already grabbed it.

**Reply by t-paul** [reply]
If you would like to see anything updated on the website part, you can do that via PR or just let me know. For the repository with the project proposals you should have direct access as all the other mentors. We'll have to see how the discussion goes regarding the presentation, but I hope we can keep the project list there. I don't think it's viable to host a separate shared application just for presenting the projects, nobody has time for that. As mentioned in the Zulip chat, I can see an additional nicer view-only front-end as entry page replacing the current very static github-page at https://opencax.github.io/.

**Comment by t-paul** [comment]
Initial update for 2024 at https://opencax.github.io/ - we still need to collect the definitive list of projects to take part this year and probably dust of some of the sub-pages.

**Reply by elalish** [reply]
When should we start putting our projects into this issue [list](https://github.com/opencax/GSoC/issues)? What do you think is a good number?

**Reply by pca006132** [reply]
I think we probably want to claim the projects first, for the potential mentor thing, and then the potential mentor can put the project into the issue list.

Should we add a communication channel? It seems that everyone else have something other than github discussion. Maybe we want to use discord/zulip? I am OK with any of them.

For project home page, we probably want to use https://github.com/elalish/manifold instead of manifoldcad page?

**Reply by elalish** [reply]
I think that's just because those projects are older than Github discussions are. I prefer to have everything in one place. ManifoldCAD does link to the github, plus it's a bit more exciting from a newbie perspective. 

@t-paul I read something about Jan 22nd - when are we trying to get things ready by?

**Reply by t-paul** [reply]
Jan 22nd the org applications open, that means Sean will apply as BRL-CAD umbrella org to GSoC, I would assume we need to have all ducks in a row when the application deadline happens at Feb 6th (https://developers.google.com/open-source/gsoc/timeline). After that the Google project admins will likely start looking at the referenced presentations.

Following what Sean said, a good number for proposals might be 3 - 5. With 3 mentors, the suggestion would be taking on at most 2 projects. Ideally every project would have 2 mentors.

As for communication channel, it seems fine to just keep Github discussions, it's indeed only usable after most of the projects found their places. When thinking about some place to chat and IRC seems not attractive enough, I'd also point to Matrix as potential option. The closed nature of Discord feels not ideal for open source projects as it makes it difficult to get data out. Even Github makes it very easy to get pretty much all data out via a backup - I do that once a month for the most central OpenSCAD repos. I have not seen that for Discord and history shows policies *will* change at some point.

**Comment by pca006132** [comment]
@elalish @zalo There are some discussions going on in the BRL-CAD zulip (https://brlcad.zulipchat.com/#narrow/stream/111975-Google-Summer-of-Code/topic/manifold), it would be better if you can join the discussion over there. Zulip provides email discussion by default so you will not miss the messages there even if you are not online.
